aten/src/ATen/BatchingRegistrations.cpp

```diff
@@ -0,0 +1,67 @@
+#include <torch/library.h>
+#include <ATen/VmapTransforms.h>
+#include <ATen/ATen.h>
+
+namespace at {
+
+// NOTE: [What is a batching rule?]
+//
+// A *batching rule* implements the logic of how to call an operator on inputs
+// that have zero or more additional batch dimensions. When one does a vmap, the
+// dimension(s) being vmap'ed over get recorded as batch dimensions.
+//
+// For example, vmap(torch.add)(x, y)
+// 1. wraps `x` into batched_x = BatchedTensor(x, bdims=[(lvl=1, dim=0)];
+// 2. wraps `y` into batched_y = BatchedTensor(y, bdims=[(lvl=1, dim=0)];
+// 3. and then runs `torch.add(batched_x, batched_y)`.
+
+// NOTE: [When should I add a batching rule?]
+// When you are adding a new operator, you'll need to add a batching rule so
+// that vmap can work efficiently with said operator. If you do not, we'll attempt
+// to generate a slow fallback for the batching rule (this is not yet implemented).
+
+// NOTE: [How to write batching rules?]
+// The signature of a batching rule should look like exactly like the C++ signature
+// of its operator.
+//
+// At a high level, what a batching rule does is the following:
+// 1. Converts (logical) BatchedTensors to views on physical tensors.
+// 2. Converts logical arguments (e.g. dimension indexes, shapes) to physical
+//    arguments that correspond to the physical tensors.
+// 3. Calls at:: operations on the physical tensors and arguments to produce
+//    some physical results.
+// 4. Converts physical results back to BatchedTensors.
+//
+// Steps 1, 2, and 4 differ for operators with different batching behaviors. When
+// writing a new batching rule, please select an VmapTransform that matches the
+// batching behavior of your operation. The VmapTransform provides helper functions
+// to do steps (1), (2), and (4).
+// (see NOTE: [What is an VmapTransform?] in VmapTransforms.h)
+
+// Note: [Future plans]
+// The API for writing a batching rule isn't stable. In the future, we'd like
+// to think about the problem of translating these batching rules to TorchScript.
+// Ideally batching rules in eager mode vs TorchScript would look pretty similar,
+// if not use the same mechanism. In order to accomplish that we might have to
+// do some refactoring.
+
+Tensor sum_batching_rule(const Tensor& self, IntArrayRef dims, bool keepdim, optional<ScalarType> dtype) {
+  auto self_physical = MultiBatchVmapTransform::logicalToPhysical(self);
```

Comment:
Reading over this, it also occurs to me that "logical" and "physical" are also very generic concepts which could apply to a variety of situations, not just vmap. So maybe we should try to find some sort of vmap-y nomenclature here. A really silly name is "batched" and "unbatched"; the "batched" naming makes sense, but unbatched doesn't really make any semantic sense, except maybe in the sense that unbatched tensors are "not" batch tensors (but they're still batched, just implicitly). After spending another thirty seconds thinking about it, I couldn't think of anything good, so maybe logical-physical is fine (since these seem to always be contextualized on one of these vmap transform things)

Reply:
I think we've overloaded the word "batch" here and that is why "batched" and "unbatched" seem weird. What we mean by a "batchedtensor" is a "vmap'ed tensor" and a "batch dim" iss a "dimension being vmap'ed over", so maybe these things should be called "VmappedTensor" and "Vmapped dim" instead and we should use "Vmaped" view and "UnVmaped" view instead of "logical" and "physical"...

Reply:
What about "representation" and "implementation"? A BatchedTensor is a presentation for a Tensor with batch dimensions, which represents a collection of Tensors of same size. The Tensor underlying a BatchedTensor is the implementation. It implements this collection of Tensors by packing them into a single Tensor of higher dimension, where some dimensions are reserved to implement batching only and not exposed to the user. It's these dimensions that we need to navigate when implementing a kernel.